common/src/main/java/org/apache/celeborn/common/network/client/TransportClientFactory.java

```diff
@@ -130,7 +136,40 @@ public TransportClientFactory(
    */
   public TransportClient createClient(String remoteHost, int remotePort, int partitionId)
       throws IOException, InterruptedException {
-    return createClient(remoteHost, remotePort, partitionId, new TransportFrameDecoder());
+    return retryCreateClient(remoteHost, remotePort, partitionId, TransportFrameDecoder::new);
+  }
+
+  public TransportClient retryCreateClient(
+      String remoteHost,
+      int remotePort,
+      int partitionId,
+      Supplier<ChannelInboundHandlerAdapter> supplier)
+      throws IOException, InterruptedException {
+    int numTries = 0;
+    while (numTries < maxClientConnectRetries) {
+      try {
+        return createClient(remoteHost, remotePort, partitionId, supplier.get());
+      } catch (Exception e) {
+        numTries++;
+        logger.warn(
+            "Retry create client, times {}/{} with error: {}",
+            numTries,
+            maxClientConnectRetries,
+            e.getMessage(),
+            e);
+        if (e instanceof InterruptedException) {
```

Comment:
> Why does this remove `IOException` check?

Only for setting the interrupt, but it will still throw an error if the maximum number of client connection retries is reached.